### PR TITLE
Support transaction policies

### PR DESCRIPTION
**Pull request by xgreenx**
Closes https://github.com/FuelLabs/fuel-vm/issues/576 Closes https://github.com/FuelLabs/fuel-core/issues/1398

Added support for transaction policies. The `Script` and `Create` 
    transactions received a new field, `policies`. Policies allow the addition 
    of some limits to the transaction to protect the user or specify some details regarding execution.

This change makes the `GasPrice` and `Maturity` fields optional, allowing to save space in the future. Also, this will enable us to support multidimensional prices later.
    
Along with this change, we introduced two new policies:
- `WitnessLimit` - allows the limitation of the maximum size of witnesses in bytes for the contract. Because of the changes in the gas calculation model(the blockchain also charges the user for the witness data), the user should protect himself from the block producer or third parties blowing up witness data and draining the user's funds.
- `MaxFee` - allows the upper bound for the maximum fee that users agree to pay for the transaction.

This change brings the following modification to the gas model:
- The `GasLimit` only limits the script execution. Previously, the `GasLimit` also limited the predicate execution time, but it is not valid anymore. So, it is not possible to use the `GasLimit` for transaction cost limitations. A new `MaxFee` policy is a way to do that. The `GasLimit` field was removed from the `Create` transaction because it only relates to the script execution(which the `Create` transaction doesn't have).
- The blockchain changes the user for the size of witness data(before it was free). There is no separate price for the storage, so it uses gas to charge the user. It affects the `min_gas` and `min_fee` calculation.
- A new `WhitnessLimit` also impacts the `max_gas` and `max_fee` calculation along with the `GasLimit`(in the case of `Create` transaction only `WitnessLimit` affects the `max_gas` and `max_fee`).
- The minimal gas also charges the user for transaction ID calculation.
      
The change has the following modification to the layout transaction:
- The `Create` transaction doesn't have the `GasLimit` field anymore. Because the `Create` transaction doesn't have any script to execute
- The `Create` and `Script` transactions don't have explicit `maturity` and `gas_price` fields. Instead, these fields can be set via a new `policies` field.
- The `Create` and `Script` transactions have a new `policies` field with a unique canonical serialization and deserialization for optimal space consumption.
    
Other breaking changes caused by the change:
- Each transaction requires setting the `GasPrice` policy.
- Previously, `GasLimit` should be less than the `MAX_GAS_PER_TX` constant. After removing this field from the `Create` transaction, it is impossible to require it. Instead, it requires that `max_gas <= MAX_GAS_PER_TX` for any transaction. Consequently, any `Script` transaction that uses `MAX_GAS_PER_TX` as a `GasLimit` will always fail because of a new rule. Setting the estimated gas usage instead solves the problem.
- If the `max_fee > policies.max_fee`, then transaction will be rejected.
- If the `witnessses_size > policies.witness_limit`, then transaction will be rejected.
- GTF opcode changed its hardcoded constants for fields. It should be updated according to the values from the specification on the Sway side.

### CODE REVIEW

File: fuel-asm/src/args.rs
Comment (MitchTurner) [inline fuel-asm/src/args.rs:87]: Why does it skip from `0x00D` to `0x100`?

Comment (MitchTurner) [reply fuel-asm/src/args.rs:87]: It looks like jumps on a few of them.

File: fuel-tx/src/tests/valid_cases/transaction.rs
Comment (MitchTurner) [inline fuel-tx/src/tests/valid_cases/transaction.rs:120]: Could we give this name more context? Like, its testing an expected failure mode, it would help of the test name communicated why this is failing.
e.g.
`check__fails_when_too_many_witnesses` or whatever. 

"Given, when, then" would also help. That might mean these should be separate tests.

File: fuel-tx/src/transaction.rs
Comment (MitchTurner) [inline fuel-tx/src/transaction.rs:562]: Curious why you include `?Sized` in all the blanket impls.

File: fuel-asm/src/args.rs
Comment (xgreenx) [reply fuel-asm/src/args.rs:87]: The values are from the [specification](https://github.com/FuelLabs/fuel-specs/pull/514/files#diff-ee3e0bbd57928f651eca1142f65942e1592743ec5fd20e20b8a44c9998679b06)

File: fuel-tx/src/transaction.rs
Comment (xgreenx) [reply fuel-tx/src/transaction.rs:562]: By default in Rust, the implementation requires `Sized`, while the trait definition - no. Because implementation of the `Policies` trait is enough to have the other four traits implementations (`GasPrice`, `WitnessLimit`, `Maturity` `MaxFeeLimit`), I don't want to require `Sized` on the `Policies` or in any other bounds which try to utilize this blanket implementation.

You can try to remove `?Sized` and you will see that I'm talking about=)

File: fuel-tx/src/transaction/policies.rs
Comment (Voxelot) [inline fuel-tx/src/transaction/policies.rs:34]: Might look better to use bitshifts or binary representation

```suggestion
        const WitnessLimit = 1<<2;
        /// If set, the maturity is present in the policies.
        const Maturity = 1<<3;
        /// If set, the max fee is present in the policies.
        const MaxFee = 1<<4;
```

File: fuel-tx/src/transaction.rs
Comment (MitchTurner) [reply fuel-tx/src/transaction.rs:562]: Hmm. Okay. At a high level I understand what it's saying.

> You can try to remove ?Sized and you will see that I'm talking about=)

It still compiles without it. So I guess I was wondering if it's a "best practice" sort of thing or if it is required by something depending on the trait.

I'm not worried about it, just trying to learn.

File: fuel-tx/src/transaction/validity.rs
Comment (xgreenx) [inline fuel-tx/src/transaction/validity.rs:349]: The specification says `GasLimit > MAX_GAS_PER_TX`, while in real life should limit the total transaction cost.

I've created a [specification PR](https://github.com/FuelLabs/fuel-specs/pull/533) to actualize it.

But it is a huge breaking change because SDK right now uses `MAX_GAS_PER_TX ` as a `GasLimit` -> the transaction always fails because of this check.

File: fuel-vm/src/interpreter/internal/message_tests.rs
Comment (xgreenx) [inline fuel-vm/src/interpreter/internal/message_tests.rs:21]: Because we removed two fields, we need to subtract `16`.

File: fuel-tx/src/transaction/fee.rs
Comment (MitchTurner) [inline fuel-tx/src/transaction/fee.rs:142]: I was thinking `witness_limit()` returned the _count_ of witnesses. You are subtracting the _size_ of all the witnesses, so maybe I was mistaken. Just want to make sure you aren't subtracting size from count.

Comment (xgreenx) [reply fuel-tx/src/transaction/fee.rs:142]: `WitnessLimit` limits the total size of witnesses in bytes=) Because each witness can be arbitrary in size, limiting by the count will not help

image

File: fuel-tx/src/tests/valid_cases/transaction.rs
Comment (MitchTurner) [inline fuel-tx/src/tests/valid_cases/transaction.rs:146]: This should probably be 2 tests. Then you'd have something like:

```
// given
let witness_size = Signature::LEN + vec![0u8; 0].size_static();

// when
let limit = witness_size;
let res = TransactionBuilder::script(generate_bytes(rng), generate_bytes(rng))
        .add_random_fee_input()
        .witness_limit(limit as u64)
        .finalize()
        .check(block_height, &test_params());
        
// then
assert!(res.is_ok());
```
And the other can just do the same but with:
```
let limit = witness_size - 1;
```
with the same `assert_eq!`

The same goes for all the following tests.

File: fuel-tx/src/transaction/policies.rs
Comment (MitchTurner) [inline fuel-tx/src/transaction/policies.rs:274]: I think these could both have better names. `works` implies that the reader knows what the desired behavior is, but one of the points of the tests is to communicate the desired behaviour:
https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#naming-your-tests

Both could benefit from some given-when-then love as well:
https://martinfowler.com/bliki/GivenWhenThen.html

File: fuel-tx/src/transaction/policies.rs
Comment (MitchTurner) [inline fuel-tx/src/transaction/policies.rs:289]: I think it's generally suggested to only have one `assert` per test. I'm kinda flexible on that when the entire test is the same except for the `then` section. That's assuming that it's communicated well why you're checking multiple things.

File: fuel-tx/src/transaction/types/create.rs
Comment (MitchTurner) [inline fuel-tx/src/transaction/types/create.rs:174]:
```suggestion
            .unwrap_or(0);
```
Helps with readability.

File: fuel-tx/src/transaction/consensus_parameters/gas.rs
Comment (Dentosal) [reply fuel-tx/src/transaction/consensus_parameters/gas.rs:589]: Adding a comment explaining that would be nice. Although, I'm also considering using `0` to make them free anyway in https://github.com/FuelLabs/fuel-vm/pull/625

File: fuel-tx/src/transaction/policies.rs
Comment (Dentosal) [inline fuel-tx/src/transaction/policies.rs:26]: Is there a particular reason to use `u32`?

File: fuel-tx/src/transaction/types/script.rs
Comment (Dentosal) [inline fuel-tx/src/transaction/types/script.rs:84]: Where does this value come from? Also using `/ 2` instead of a bit shift would be cleaner anyway.

File: fuel-vm/src/interpreter/memory/tests.rs
Comment (Dentosal) [inline fuel-vm/src/interpreter/memory/tests.rs:29]: Needs explanation.

File: fuel-vm/src/interpreter/metadata.rs
Comment (Dentosal) [reply fuel-vm/src/interpreter/metadata.rs:167]: Just throw an error. Returning an invalid value is not backward compability, it's just wrong.

File: fuel-vm/src/tests/predicate.rs
Comment (Dentosal) [inline fuel-vm/src/tests/predicate.rs:145]: Do we need this print here?

File: fuel-vm/src/interpreter/memory/tests.rs
Comment (MitchTurner) [inline fuel-vm/src/interpreter/memory/tests.rs:29]: What's going on here? Isn't this just halving the `max_gas_per_tx`?

File: fuel-vm/src/util.rs
Comment (MitchTurner) [inline fuel-vm/src/util.rs:594]: Here too.

File: fuel-tx/src/builder.rs
Comment (xgreenx) [reply fuel-tx/src/builder.rs:122]: Because of this check:
image

We use builders mostly for tests and it is better to set this policy for all tests instead of updating all places.

File: fuel-tx/src/transaction/policies.rs
Comment (xgreenx) [reply fuel-tx/src/transaction/policies.rs:26]: https://github.com/FuelLabs/fuel-specs/pull/514#discussion_r1366359260

File: CHANGELOG.md
Comment (Voxelot) [inline CHANGELOG.md:25]:
```suggestion
    - The blockchain charges the user for the size of witness data(before it was free). There is no separate price for the storage, so it uses gas to charge the user. It affects the `min_gas` and `min_fee` calculation.
```

File: CHANGELOG.md
Comment (Voxelot) [inline CHANGELOG.md:26]:
```suggestion
    - A new policy called `WitnessLimit` also impacts the `max_gas` and `max_fee` calculation in addition to `GasLimit`(in the case of `Create` transaction only `WitnessLimit` affects the `max_gas` and `max_fee`).
```

File: CHANGELOG.md
Comment (Voxelot) [inline CHANGELOG.md:29]:
```suggestion
    The change has the following modification to the transaction layout:
```

**Review by Dentosal — DISMISSED** [review]
LGTM; I suggest you wait for another review as well.

File: CHANGELOG.md
Comment (Voxelot) [inline CHANGELOG.md:19]:
```suggestion
    `GasLimit` was renamed to `ScriptGasLimit`.
```

File: CHANGELOG.md
Comment (Voxelot) [inline CHANGELOG.md:26]:
```suggestion
    - The `ScriptGasLimit` only limits script execution. Previously, the `ScriptGasLimit` also limited the predicate execution time, instead predicate gas is now directly included into `min_fee`. So, it is not possible to use the `ScriptGasLimit` for transaction cost limitations. A new `MaxFee` policy is a way to do that. The `GasLimit` field was removed from the `Create` transaction because it only relates to the script execution (which the `Create` transaction doesn't have).
```

File: fuel-asm/src/args.rs
Comment (Voxelot) [inline fuel-asm/src/args.rs:251]:
```suggestion
        /// Set `$rA` to `tx.policyTypes`
```

File: fuel-asm/src/args.rs
Comment (Voxelot) [inline fuel-asm/src/args.rs:252]: We should bump these up since we aren't using `GTF_POLICY_COUNT` anymore

```suggestion
        PolicyTypes = 0x500,
```

File: fuel-tx/src/builder.rs
Comment (Voxelot) [inline fuel-tx/src/builder.rs:300]:
```suggestion
    pub fn script_gas_limit(&mut self, gas_limit: Word) -> &mut Self
```

File: fuel-tx/src/tests/valid_cases/transaction.rs
Comment (Voxelot) [inline fuel-tx/src/tests/valid_cases/transaction.rs:172]:
```suggestion
fn script_set_witness_limit_less_than_witness_data_size_fails() {
```

File: fuel-tx/src/transaction.rs
Comment (Voxelot) [inline fuel-tx/src/transaction.rs:42]: Should pub mods be sorted into a separate section from private mods?

File: fuel-vm/src/tests/encoding.rs
Comment (MitchTurner) [reply fuel-vm/src/tests/encoding.rs:220]: Gonna bump this. 

I don't think we should be using `>>` and I think it would be best to move the values to variables.

File: fuel-tx/src/transaction/policies.rs
Comment (xgreenx) [reply fuel-tx/src/transaction/policies.rs:64]: The idea was to write the implementation manually to control values because, in the future, we may remove `GasPrice` or any other policy.

**Review by MitchTurner — DISMISSED** [review]

File: fuel-tx/src/transaction/policies.rs
Comment (bvrooman) [reply fuel-tx/src/transaction/policies.rs:64]: As long as they don't get out of sync 👍

**Comment by bvrooman** [comment]
I proposed some changes in https://github.com/FuelLabs/fuel-vm/pull/630 - mostly for the purpose of allowing us to encapsulate policy definitions and checks better. It would be nice to have this logic centralized rather than spread out, because it means a reduction in overall complexity: Having one place to define this logic would mean less duplication and easier extensibility. But I am simply suggesting one approach - if you have an idea on how you would like to do this, we can review that too now or at a later date after this PR is in.

**Review by bvrooman — APPROVED** [review]

**Pull request merged**
